routemaster/services/autodrop.rb

```diff
@@ -0,0 +1,39 @@
+require 'routemaster/services'
+require 'routemaster/models/database'
+require 'routemaster/models/subscription'
+require 'routemaster/mixins/log'
+require 'ostruct'
+
+module Routemaster
+  module Services
+    # Autodrop messages from queues until the database is empty enough.
+    #
+    # Naive algorithm for now: just drop the oldest messages from all queues
+    # until we're below the low-water mark.
+    class Autodrop
+      include Mixins::Log
+
+      BATCH_SIZE = 100
+
+      def initialize
+        @database = Models::Database.instance
+      end
+
+      def call
+        _log.info { 'auto-drop: starting' }
+        return false unless @database.too_full?
+        messages_removed = 0
+        queues = Models::Subscription.map(&:queue)
+
+        # loop through queues, removing messages
+        until @database.empty_enough?
```

Comment:
Of the whole loop, since we check for the size, sort them and then iterate again to drop them. It's probably premature optimisation and this shouldn't happen often but just curious.